storage/client/bucket.cc

```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/bucket.h"
+#include "google/cloud/internal/throw_delegate.h"
+#include <sstream>
+#include <thread>
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+static_assert(std::is_copy_constructible<storage::Bucket>::value,
+              "storage::Bucket must be constructible");
+static_assert(std::is_copy_assignable<storage::Bucket>::value,
+              "storage::Bucket must be assignable");
+
+BucketMetadata Bucket::GetMetadata() {
+  // TODO(#555) - use policies to implement retry loop.
+  Status last_status;
+  for (int i = 0; i != 3; ++i) {
+    auto result = client_->GetBucketMetadata(bucket_name_);
+    last_status = std::move(result.first);
+    if (last_status.ok()) {
+      return std::move(result.second);
+    }
+    // TODO(#581) - use policies to determine what error codes are permanent.
+    if (503 != last_status.status_code()) {
+      // For the moment, only SERVICE UNAVAILABLE (503) is retried.
+      std::ostringstream os;
+      os << "Permanent error in Bucket::GetMetadata: " << last_status;
+      google::cloud::internal::RaiseRuntimeError(os.str());
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  std::ostringstream os;
+  os << "Retry policy exhausted in Bucket::GetMetadata: " << last_status;
+  google::cloud::internal::RaiseRuntimeError(os.str());
+}
+
+void Bucket::ValidateBucketName(std::string const& bucket_name) {
```

Comment:
In general, we should avoid checking on the client side, the server is always more accurate, and it would be very sad if the library decided to reject a name that was actually valid.  I did it here because then I do not need to URL encode the name.

Do you want me to change this in this PR?

And just wondering: presumably Google can create buckets that start with "goog" and applications should be able to access them (assuming right permissions and so forth)?

Reply:
I was thinking about validation in the context of creating buckets; accessing such buckets should be allowed, you're correct.

Looking at the other client libraries [1], the Python lib only checks one restriction [2], and the Node lib doesn't check any of them... but it does apparently silently "fix" the name for users if they include a gsutil-style naming pattern (e.g. "gs://bucket" --> "bucket") [3].  I really don't like that precedent :(  Anyway, most of the libs do no (or next to no) validation, which I agree with -- all we should check for is whether the user supplied a non-empty string.

[1] https://cloud.google.com/apis/docs/cloud-client-libraries
[2] https://github.com/GoogleCloudPlatform/google-cloud-python/blob/b1620bf734422cedbab4527ec1e7fcea486196e7/storage/google/cloud/storage/_helpers.py#L24
[3] https://github.com/googleapis/nodejs-storage/blob/8110cceeb37cdd01dd90389ce1a4a979b0bfbc27/src/bucket.js#L63